### DISCUSSION TITLE
Support for object cloning ?

### DISCUSSION THREAD

**Discussion by didier31** [LogtalkDotOrg/logtalk3#136] [Q&A]
How to clone a prototype or an instance in Logtalk ?
Does it exist support for this operation ?

**Comment by pmoura** [comment]
There's a `cloning` protocol in the library but no implementation for it as the details are usually application specific. For example, if an object contains references to other objects, some applications may require a *shallow* cloning while others a *deep* cloning. Same issues arise when implementing object serialization.

My usual advice here, assuming the object plays the role of a prototype, is to extend the object instead and then define any predicates that make changes to the object internal dynamic state (which is implicit in this case, otherwise there would be no need for cloning) so that the changes are local to the object receiving the message. Therefore, when the descendant object changes its  internal dynamic state, that change will not affect the parent object. A relevant example is:

https://github.com/LogtalkDotOrg/logtalk3/tree/master/examples/design_patterns/creational/prototype

When using objects playing the roles of classes and instances, two relevant examples are:

https://github.com/LogtalkDotOrg/logtalk3/tree/master/examples/classvars
https://github.com/LogtalkDotOrg/logtalk3/tree/master/examples/instvars

You can of course also define an implementation for the `cloning` protocol that queries an object dynamic state and then set the same state in a new object with both objects sharing an ancestor (object or category) for their static part.
